feat(immutable-arraybuffer,ses): drop the pseudo-prototype intrinsic (per DESIGN.md) (#435)
endojs/endo-but-for-bots: packages/immutable-arraybuffer/README.md

# `@endo/immutable-arraybuffer`

This `@endo/immutable-arraybuffer` package provides a lib layer and a shim for a proposed new JavaScript feature: *Immutable ArrayBuffers*.

> - The lib layer in `src/lib.js` defines a property record of the methods and accessors the proposal adds to `ArrayBuffer.prototype`, along with the factory that constructs emulated immutable buffers.

Reviewer: @kriscendobot ,
The lib layer is no longer exposed outside this package. Therefore it should not even be mentioned in the `README.md` since that is meant for customers of this package. The lib layer is mentioned in the `DESIGN.md`, changeset, and PR comment, as it should be.

> The package re-exports `isBufferImmutable` (a brand check usable without forcing the shim to be installed) and, pending the premise-2 follow-up that narrows the package's `exports`, the `sliceBufferToImmutable` and `optTransferBufferToImmutable` free-function helpers that pre-shim callers still rely on.

Reviewer: @kriscendobot ,
The package should stop exporting `isBufferImmutable` or any other names. It should be a side-effect-only package via its one `shim.js` export.

- A shim modifies the existing JavaScript primordials as needed to most closely emulate the feature as proposed.

> The `shim.js` file copies the lib layer's property record onto `ArrayBuffer.prototype`.

Reviewer: @kriscendobot ,
Below the level of abstraction that the README.md should mention.

For a `TypedArray` object, the contents of the backing store appear as indexed data properties of the `TypeArray` object that reflect the current contents of this backing store.
Currently, because there is no way to prevent the contents of the backing store from being changed, `TypedArray`s cannot be frozen.

Some JavaScript implementations, like Moddable XS, bring JavaScript to embedded systems, like device controllers, where ROM is much more plentiful and cheaper than RAM.
These systems need to place voluminous fixed data into ROM, and currently do so using semantics outside the official JavaScript standard.

The [OCapN](https://ocapn.org/) network protocol treats strings and byte-arrays as distinct forms of bulk data to be transmitted by copy.
At JavaScript endpoints speaking OCapN such as `@endo/pass-style` + `@endo/marshal`, JavaScript strings represent OCapN strings.
The immutability of strings in the JavaScript language reflects their by-copy nature in the protocol.
Likewise, to reflect an OCapN byte-array well into the JavaScript language, we need an immutable container of bulk binary data.
There currently are none.

> An Immutable `ArrayBuffer` would provide exactly the low-level machinery we need.

Reviewer: @kriscendobot ,
The old design was that an Immutable ArrayBuffer itself would be the JS representation of an OCapN byteArray. The new design is to use a frozen Uint8Array instead.

```suggestion
A frozen `Uint8Array` would provide exactly the low-level machinery we need.
```

## The Lib Layer

The lib layer in `src/lib.js` does not modify `ArrayBuffer.prototype` directly.
Instead, it exports
- `isBufferImmutable(buffer: ArrayBuffer) :boolean` (via `index.js`) -- a brand check that returns true for emulated immutable buffers and false otherwise.
- `immutableArrayBufferLibProperties` -- a plain record whose own properties are the methods and accessors the shim installs onto `ArrayBuffer.prototype`.
This record is internal to the package; only the shim consumes it.

In order for emulated immutable buffers to be of type `ArrayBuffer`, they cannot be actual `ArrayBuffer` exotic objects (which would be writable).
Instead, an emulated immutable buffer is a plain object whose direct prototype is `ArrayBuffer.prototype`.
So `x instanceof ArrayBuffer` will act as proposed.
The lib layer maintains a brand `WeakMap` from emulated immutable buffers to the underlying genuine `ArrayBuffer` whose contents it encapsulates.
The methods the shim installs onto `ArrayBuffer.prototype` discriminate on brand membership: they treat the receiver as immutable when it is in the WeakMap, and as a genuine ArrayBuffer otherwise (the amplifier-with-this-fallthrough pattern).

> This way, one set of methods on the shared `ArrayBuffer.prototype` correctly handles both genuine and emulated buffers.

Reviewer: @kriscendobot ,
The README.md should not mention the lib layer. If the text here makes any points not made in DESIGN.md, those points should be moved there.

For genuine ArrayBuffers, the overwritten methods delegate to the captured genuine method and behave identically to before.
For emulated immutable buffers, the methods either return the appropriate immutable behaviour (for `slice`) or throw the appropriate "cannot mutate" `TypeError` (for the mutators).

A warning fires if the shim is about to overwrite a property of `ArrayBuffer.prototype` that was not already expected to be overwritten.
The four genuine mutator-method overwrites (`slice`, `resize`, `transfer`, `transferToFixedLength`) are on a static expected-overwrite list and do not trigger the warning; any other overwrite indicates a platform that has independently shipped part of the proposal's surface, which the shim author should investigate.

## Caveats

The *Immutable ArrayBuffer* shim falls short of the proposal in the following ways
- The lib layer and shim rely on the underlying platform having either `structuredClone` or `ArrayBuffer.prototype.transfer`.
However, Node <= 16 has neither.
Node 17 introduces `structuredClone` and Node 21 introduces `ArrayBuffer.prototype.transfer`.

> Without either, the lib layer and shim fail to initialize.

Reviewer: @kriscendobot ,
No longer true.

```suggestion
Without either, the shim still shims `ArrayBuffer.prototype.sliceToImmutable` but omits `ArrayBuffer.prototype.transferToImmutable`. Thus, even after importing the shim, code may want to feature test for `ArrayBuffer.prototype.transferToImmutable`.
```

- The shim's emulated immutable buffers are not real `ArrayBuffer` exotic objects.
If they were, the shim would not be able to protect them from being written.
Even though they implement the full proposed `ArrayBuffer` API, they cannot be plug-compatible: they cannot be used as the backing stores of `DataView`s or `TypedArray`s.
Perhaps follow-on shims might modify `DataView` and `TypedArray` to emulate that as well, but that is hard and beyond the ambition of this lib + shim.
- Unlike genuine `ArrayBuffer` or `SharedArrayBuffer` exotic objects, the shim's emulated immutable buffers cannot be cloned or transfered between JS threads.
- Even after the *Immutable ArrayBuffer* proposal is implemented by the platform, the current code will still replace it with the shim implementation, in accord with shim best practices.

> It will require a later manual step to delete the shim, after manual analysis of the compat implications.

Reviewer: @kriscendobot ,
Since the proposal has reached stage 3, as of this PR, this is no longer true. Instead, the shim will defer to any prior implementation.

Thus, the objects and function it creates are not hardened by this shim itself.
Rather, the ses-shim is expected to import this, and then treat the resulting objects as if they were additional primordials, to be hardened during `lockdown`'s harden phase.

> ## Platform support for `transferToImmutable`

Reviewer: @kriscendobot ,
The tables in this section are great, thanks!

Can you gather enough info on Moddable XS to add to the table? If so, please do.

attn @phoddie

Reviewer 2: XS has had both `ArrayBuffer.prototype.transferToImmutable()` and `structuredClone()` for some time.

Reviewer: Roughly when? Just so it can fit smoothly into the table?

Reviewer 2: structureClone was 4 years back, according to Git. Immutable stuff was when we got Stage 1.

Reviewer: @kriscendobot please add XS according to the info above from @phoddie . It is ok if approximate.
